### PR TITLE
Reduce allocations, simplifed code, faster implementation, included tests - StreamInfo.ToUrl

### PR DESCRIPTION
**Pull request by BaronGreenback**
Method is used extensively in the device profile sub-system.

**Original code (ToUrl / BuildParam methods) :-**

Added values to a name/pair list, creating a new object for each entry. (checking and storing string.Empty for null)
parsed this list and ignored any string.Empty or default values and built a list<string> instance of individual name=value
joined the list<string> together with &
appending it at the end of the url
returned the url.

**Replacement code**
creates a string builder 2000 bytes long.
adds the url to string builder
adds the non null/default values to the string builder
returns the url

As it's a basic function re-write, included a testing unit to ensure the output remains the same.

### CODE REVIEW

File: tests/Jellyfin.Model.Tests/Dlna/StreamInfo_tests.cs [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+                    property.SetValue(destination, RandomArray(random, type.GetElementType()));
+                    continue;
+                }
+            }
+        }
+
+        [InlineData(DlnaProfileType.Audio)]
+        [InlineData(DlnaProfileType.Video)]
+        [InlineData(DlnaProfileType.Photo)]
+
```
Comment (crobibero) [inline tests/Jellyfin.Model.Tests/Dlna/StreamInfo_tests.cs:211]:
```suggestion
```

**Review by crobibero — APPROVED** [review]

File: MediaBrowser.Model/Dlna/StreamInfo.cs [hunk 2]
```diff
@@ -597,212 +598,279 @@ public string GetOption(string name)
...
-                    return string.Format(CultureInfo.InvariantCulture, "{0}/audio/{1}/master.m3u8?{2}", baseUrl, ItemId, queryString);
-                }
+                sb.Append("&Static=true");
+            }
 
-                return string.Format(CultureInfo.InvariantCulture, "{0}/audio/{1}/stream{2}?{3}", baseUrl, ItemId, extension, queryString);
+            if (VideoCodecs.Length != 0)
+            {
+                sb.Append("&VideoCodec=");
+                sb.Append(string.Join(',', VideoCodecs));
```
Comment (cvium) [inline MediaBrowser.Model/Dlna/StreamInfo.cs:677]:
```suggestion
                sb.AppendJoin(',', VideoCodecs);
```

File: MediaBrowser.Model/Dlna/StreamInfo.cs [hunk 3]
```diff
@@ -597,212 +598,279 @@ public string GetOption(string name)
...
             }
 
-            if (string.Equals(SubProtocol, "hls", StringComparison.OrdinalIgnoreCase))
+            if (AudioCodecs.Length != 0)
             {
-                return string.Format(CultureInfo.InvariantCulture, "{0}/videos/{1}/master.m3u8?{2}", baseUrl, ItemId, queryString);
+                sb.Append("&AudioCodec=");
+                sb.Append(string.Join(',', AudioCodecs));
```
Comment (cvium) [inline MediaBrowser.Model/Dlna/StreamInfo.cs:683]:
```suggestion
                sb.AppendJoin(',', AudioCodecs);
```

File: MediaBrowser.Model/Dlna/StreamInfo.cs [hunk 4]
```diff
@@ -597,212 +598,279 @@ public string GetOption(string name)
...
-                }
+                sb.Append("&Tag=");
+                sb.Append(etag);
+            }
 
-                list.Add(new NameValuePair("BreakOnNonKeyFrames", item.BreakOnNonKeyFrames.ToString(CultureInfo.InvariantCulture)));
+            if (SubtitleStreamIndex.HasValue && SubtitleDeliveryMethod == SubtitleDeliveryMethod.Embed && SubtitleCodecs.Length != 0)
+            {
+                sb.Append("&SubtitleCodec=");
+                sb.Append(string.Join(',', SubtitleCodecs));
```
Comment (cvium) [inline MediaBrowser.Model/Dlna/StreamInfo.cs:845]:
```suggestion
                sb.AppendJoin(',', SubtitleCodecs);
```

File: MediaBrowser.Model/Dlna/StreamInfo.cs [hunk 5]
```diff
@@ -597,212 +598,279 @@ public string GetOption(string name)
...
+                sb.Append('=');
+                sb.Append(pair.Value.Replace(" ", string.Empty, StringComparison.Ordinal));
             }
 
-            if (!item.IsDirectStream)
+            if (!IsDirectStream && TranscodeReasons.Length > 0)
             {
-                list.Add(new NameValuePair("TranscodeReasons", string.Join(',', item.TranscodeReasons.Distinct())));
+                sb.Append("&TranscodeReasons=");
+                sb.Append(string.Join(',', TranscodeReasons.Distinct().Select(i => i.ToString())));
```
Comment (cvium) [inline MediaBrowser.Model/Dlna/StreamInfo.cs:865]:
```suggestion
                sb.AppendJoin(',', TranscodeReasons.Distinct().Select(i => i.ToString()));
```

**Comment by Shadowghost** [comment]
Superseded by #9369

**Pull request closed without merging**